cirq-core/cirq/circuits/moment.py

```diff
@@ -669,25 +676,92 @@ def _commutes_(self, other: Any, *, atol: float = 1e-8) -> Union[bool, NotImplem
 
         Returns:
             True: The Moment and Operation commute OR they don't have shared
-            quibits.
+                qubits.
             False: The two values do not commute.
             NotImplemented: In case we don't know how to check this, e.g.
                 the parameter type is not supported yet.
         """
-        if not isinstance(other, ops.Operation):
-            return NotImplemented
-
-        other_qubits = set(other.qubits)
-        for op in self.operations:
-            if not other_qubits.intersection(set(op.qubits)):
-                continue
-
-            commutes = protocols.commutes(op, other, atol=atol, default=NotImplemented)
+        if isinstance(other, ops.Operation):
+            # If an Operation is provided, convert this to a Moment consisting only
+            # of the given Operation
+            return self._commutes_(Moment(other), atol=atol)
+
+        if isinstance(other, Moment):
+            # Check if sets of qubits overlap. If not, then no need to go any further.
+            if not set(self.qubits) & set(other.qubits):
+                return True
+
+            # Check pairwise commuting between all pairs of
+            # operations. If they all commute then no
+            # need to go any further
+            if all(
+                definitely_commutes(op_1, op_2, atol=atol)
+                for op_1, op_2 in itertools.product(self.operations, other.operations)
+            ):
+                return True
+
+            # Decompose into disjoint overlapping sets of qubits
+            qubit_subsets = [list(op.qubits) for op in self.operations + other.operations]
+            disjoint_set = DisjointSet(itertools.chain.from_iterable(qubit_subsets))
+            for subset in qubit_subsets:
```

Comment:
what happens if a qubit appears in both moments in operations with different qubits?
so in m1: op1 acting on (q1, q0) in m2 op2 acting on q3, q4, q0)